tmt/utils.py

```diff
@@ -276,41 +347,74 @@ def _log(self, message):
+            err: bool = False) -> None:
         """ Print a message regardless the quiet mode """
         self._log(self._indent(key, value, color=None, shift=shift))
         echo(self._indent(key, value, color, shift), err=err)
 
-    def info(self, key, value=None, color=None, shift=0, err=False):
+    def info(
+            self,
+            key: str,
+            value: Optional[str] = None,
+            color: Optional[str] = None,
+            shift: int = 0,
+            level: int = 1,
+            err: bool = False) -> None:
         """ Show a message unless in quiet mode """
         self._log(self._indent(key, value, color=None, shift=shift))
         if not self.opt('quiet'):
             echo(self._indent(key, value, color, shift), err=err)
 
-    def warn(self, message, shift=0):
-        """ Show a yellow warning message on info level, send to stderr """
-        self.info('warn', message, color='yellow', shift=shift, err=True)
-
-    def fail(self, message, shift=0):
-        """ Show a red failure message on info level, send to stderr """
-        self.info('fail', message, color='red', shift=shift, err=True)
-
     def verbose(
-            self, key, value=None, color=None, shift=0, level=1, err=False):
+            self,
+            key: str,
+            value: Optional[str] = None,
+            color: Optional[str] = None,
+            shift: int = 0,
+            level: int = 1,
+            err: bool = False) -> None:
         """ Show message if in requested verbose mode level """
         self._log(self._indent(key, value, color=None, shift=shift))
         if self.opt('verbose') >= level:
             echo(self._indent(key, value, color, shift), err=err)
 
-    def debug(self, key, value=None, color=None, shift=1, level=1, err=False):
+    def debug(
+            self,
+            key: str,
+            value: Optional[str] = None,
+            color: Optional[str] = None,
+            shift: int = 0,
+            level: int = 1,
+            err: bool = False) -> None:
         """ Show message if in requested debug mode level """
         self._log(self._indent(key, value, color=None, shift=shift))
         if self.opt('debug') >= level:
             echo(self._indent(key, value, color, shift), err=err)
 
-    def _run(
-            self, command, cwd, shell, env, log, join=False, interactive=False,
-            timeout=None):
+    def warn(self, message: str, shift: int = 0) -> None:
+        """ Show a yellow warning message on info level, send to stderr """
+        self.info('warn', message, color='yellow', shift=shift, err=True)
+
+    def fail(self, message: str, shift: int = 0) -> None:
+        """ Show a red failure message on info level, send to stderr """
+        self.info('fail', message, color='red', shift=shift, err=True)
+
+    def _run(self,
+             command: Union[str, List[str]],
+             cwd: Optional[str],
+             shell: bool,
+             env: Optional[ContextType],
```

Comment:
Of course, it's possible to do the simplest change, `EnvironmentType = ContextType` at the top of the file, if they are the same thing. On the other hand, if they are the same type, wouldn't one type only make things easier to reason about?

Reply:
Hm, this seems to be rather a matter of personal taste. I'd vote for two explicit names even if they are "duplicating" the same type. Anyway, thinking about this a bit more I realized the `ContextType` actually should have a list of strings as values:

    ContextType = Dict[str, List[str]]

Because that's what `context_to_dict()` is producing when processing `--context` command line options.

Reply:
I've found a single use of `context_to_dict()`, `obj.fmf_context = tmt.utils.context_to_dict(context)` in `cli.py` - would you mind calling the type `FmfContextType`, to make it clear it's not just "random" context but indeed something tied with FMF and its concept of context?

Couple with that, I'd add the `EnvironmentType` which would be the `Dict[str, str]`, to pass around things like envvars & similar data. Would this be aligned with you view?

Reply:
Yeah, `FmfContextType` makes a perfect sense! Thanks.

Reply:
Done, in 4113285ba9050dd98fe04d5d674fed7425d6ba25 I added `FmfContextType` and `EnvironmentType`, let me know whether it's what you expected.

Reply:
Great, thanks!